File: frontend/lib/pricing.ts
// Pricing bridge client. Wraps POST /api/pricing/quote.
// Gating failures (409) and engine outage (503) are turned into a typed
// "unavailable" instead of throwing so the tile can render a dash + reason.

import { api, ApiError } from "./api";

export interface PriceQuote {
  status: string;
  venue_id: string;
  party_size: number;
  party_size_category: "small" | "large";
  currency: string;
  minor_units: number;
  price_minor: number | null;
  price_major: number | null;
  base_minor: number | null;
  floor_minor: number | null;
  ceiling_minor: number | null;
  raw_minor: number | null;
  multipliers: Record<string, number>;
  predicted_wait_mins: number | null;
  premium_share_pct: number | null;
  queue_count: number | null;
  premium_count_in_category: number | null;
  hard_cap_for_category: number | null;
  session_id: string | null;
  valid_until_ts: number | null;
  valid_for_seconds: number | null;
  message: string | null;
}

export interface PriceQuoteRequest {
  party_size: number;
  service_id?: string | null;
  session_id?: string | null;
}

export interface PriceUnavailable {
  ok: false;
  reason: string;
  message: string;
}
export interface PriceOk {
  ok: true;
  quote: PriceQuote;
}
export type PriceResponse = PriceOk | PriceUnavailable;

function extractDetail(err: ApiError): { reason: string; message: string } {
  // Backend raises HTTPException(detail={reason, message}); the wrapper puts
  // the parsed JSON body in err.body. Fall back to err.message for stringy
  // details (older endpoints).
  const body = err.body as { detail?: unknown } | undefined;
  const detail = body?.detail;
  if (detail && typeof detail === "object") {
    const d = detail as { reason?: string; message?: string };
    return {
      reason: d.reason ?? "unavailable",
      message: d.message ?? "Price unavailable.",
    };
  }
  return { reason: "unavailable", message: String(err.message ?? "Price unavailable.") };
}

export const pricingApi = {
  quote: async (token: string, body: PriceQuoteRequest): Promise<PriceResponse> => {
    try {
      const quote = await api<PriceQuote>("/api/pricing/quote", {
        method: "POST",
        body,
        token,
      });
      return { ok: true, quote };
    } catch (err) {
      if (err instanceof ApiError) {
        const { reason, message } = extractDetail(err);
        return { ok: false, reason, message };
      }
      return { ok: false, reason: "network", message: "Could not reach pricing." };
    }
  },
};

export function formatPrice(q: PriceQuote): string {
  if (q.price_major == null) return "—";
  if (q.currency === "JPY") return `¥${Math.round(q.price_major).toLocaleString()}`;
  return new Intl.NumberFormat(undefined, {
    style: "currency",
    currency: q.currency,
  }).format(q.price_major);
}

File: backend/app/schemas/pricing.py
"""Schemas for the pricing bridge. PriceQuote mirrors pricing_engine.PriceResult
so the engine's JSON validates straight through. The browser sends a small
input (party_size + optional session_id / service_id); the server builds the
venue_config from the operator's own Restaurant + VenueSettings."""

from __future__ import annotations

from typing import Any, Optional

from pydantic import BaseModel, Field


class PriceQuoteRequest(BaseModel):
    party_size: int = Field(ge=1, le=20)
    service_id: Optional[str] = None
    session_id: Optional[str] = None  # pass to lock the price for 5 min


class PriceQuote(BaseModel):
    status: str
    venue_id: str
    party_size: int
    party_size_category: str
    currency: str
    minor_units: int
    price_minor: Optional[int] = None
    price_major: Optional[float] = None
    base_minor: Optional[int] = None
    floor_minor: Optional[int] = None
    ceiling_minor: Optional[int] = None
    raw_minor: Optional[int] = None
    multipliers: dict[str, Any] = {}
    predicted_wait_mins: Optional[float] = None
    premium_share_pct: Optional[float] = None
    queue_count: Optional[int] = None
    premium_count_in_category: Optional[int] = None
    hard_cap_for_category: Optional[int] = None
    session_id: Optional[str] = None
    valid_until_ts: Optional[float] = None
    valid_for_seconds: Optional[int] = None
    message: Optional[str] = None
